File: miniserve/cache/block_table.py
"""Per-request mapping from logical token positions to physical KV slots.

Token ``pos`` of a request lives in physical block ``blocks[pos // block_size]``
at offset ``pos % block_size``; its slot in a pool flattened to
``[num_blocks * block_size]`` is ``slot(pos)``.

A table may start with blocks shared with other holders (a cached prefix).
Only full blocks can be shared, so every token appended afterwards goes into a
block this table holds exclusively and no copy-on-write is needed.
"""

from __future__ import annotations

from collections.abc import Sequence

from miniserve.cache.block_allocator import BlockAllocator


class BlockTable:
    def __init__(self, allocator: BlockAllocator, prefix_blocks: Sequence[int] = ()):
        self.allocator = allocator
        self.block_size = allocator.block_size
        if len(set(prefix_blocks)) != len(prefix_blocks):
            raise ValueError(f"duplicate block in prefix {list(prefix_blocks)}")
        allocator.incref(prefix_blocks)
        self.blocks: list[int] = list(prefix_blocks)
        self.num_tokens = len(self.blocks) * self.block_size

    @property
    def capacity(self) -> int:
        return len(self.blocks) * self.block_size

    @property
    def last_block_len(self) -> int:
        """Tokens in the last block (1..block_size), or 0 for an empty table."""
        if self.num_tokens == 0:
            return 0
        return self.num_tokens - (len(self.blocks) - 1) * self.block_size

    def blocks_needed(self, n: int) -> int:
        """Additional blocks required to append ``n`` tokens."""
        if n < 0:
            raise ValueError(f"cannot append {n} tokens")
        return max(0, -(-(self.num_tokens + n) // self.block_size) - len(self.blocks))

    def append_tokens(self, n: int) -> list[int]:
        """Reserve room for ``n`` more tokens, allocating blocks as needed. All or nothing.

        Returns the newly allocated blocks.
        """
        new = self.allocator.allocate(self.blocks_needed(n))
        self.blocks += new
        self.num_tokens += n
        return new

    @property
    def num_blocks(self) -> int:
        """Blocks the table holds. Cheaper than ``len(table.blocks)``, which a
        C++ backend would have to build a list for."""
        return len(self.blocks)

    def slot(self, pos: int) -> int:
        if not 0 <= pos < self.num_tokens:
            raise IndexError(f"position {pos} out of range [0, {self.num_tokens})")
        return self.blocks[pos // self.block_size] * self.block_size + pos % self.block_size

    def rewind(self, n: int) -> None:
        """Undo the append of the last ``n`` tokens, keeping the blocks.

        The blocks stay because the caller is about to recompute those tokens
        into the same slots; freeing and reallocating would hand out different ones.
        """
        if not 0 <= n <= self.num_tokens:
            raise IndexError(f"cannot rewind {n} of {self.num_tokens} tokens")
        self.num_tokens -= n

    def tail_slots(self, n: int) -> list[int]:
        """Slots of the last ``n`` tokens, in order: one call per request per step
        instead of one per token."""
        if not 0 <= n <= self.num_tokens:
            raise IndexError(f"tail of {n} tokens out of range [0, {self.num_tokens}]")
        bs = self.block_size
        return [self.blocks[p // bs] * bs + p % bs for p in range(self.num_tokens - n, self.num_tokens)]

    def release(self) -> None:
        """Give every block back (dropping this table's reference) and empty the table."""
        self.allocator.free(self.blocks)
        self.blocks = []
        self.num_tokens = 0

File: minicore/src/block_table.cpp
#include "minicore/block_table.hpp"

namespace minicore {

BlockTable::BlockTable(BlockAllocator* allocator, const std::vector<BlockId>& prefix_blocks)
    : allocator_(allocator) {
  std::unordered_set<BlockId> seen(prefix_blocks.size() * 2);
  for (BlockId b : prefix_blocks) {
    if (!seen.insert(b).second) {
      std::string listed = "[";
      for (std::size_t i = 0; i < prefix_blocks.size(); ++i) {
        if (i) listed += ", ";
        listed += std::to_string(prefix_blocks[i]);
      }
      throw std::invalid_argument("duplicate block in prefix " + listed + "]");
    }
  }
  allocator_->incref(prefix_blocks);
  blocks_ = prefix_blocks;
  num_tokens_ = static_cast<int>(blocks_.size()) * allocator_->block_size();
}

int BlockTable::blocks_needed(int n) const {
  if (n < 0) throw std::invalid_argument("cannot append " + std::to_string(n) + " tokens");
  const int bs = allocator_->block_size();
  const int need = (num_tokens_ + n + bs - 1) / bs - static_cast<int>(blocks_.size());
  return need > 0 ? need : 0;
}

std::vector<BlockId> BlockTable::append_tokens(int n) {
  const int need = blocks_needed(n);
  const std::size_t before = blocks_.size();
  allocator_->allocate_into(need, blocks_);  // throws before touching blocks_ if short
  num_tokens_ += n;
  return std::vector<BlockId>(blocks_.begin() + static_cast<std::ptrdiff_t>(before), blocks_.end());
}

std::vector<std::int64_t> BlockTable::tail_slots(int n) const {
  if (n < 0 || n > num_tokens_) {
    throw BlockIndexError("tail of " + std::to_string(n) + " tokens out of range [0, " +
                          std::to_string(num_tokens_) + "]");
  }
  std::vector<std::int64_t> out;
  out.reserve(static_cast<std::size_t>(n));
  const int bs = allocator_->block_size();
  for (int pos = num_tokens_ - n; pos < num_tokens_; ++pos) {
    out.push_back(static_cast<std::int64_t>(blocks_[static_cast<std::size_t>(pos / bs)]) * bs +
                  pos % bs);
  }
  return out;
}

void BlockTable::release() {
  allocator_->free(blocks_);
  blocks_.clear();
  num_tokens_ = 0;
}

}  // namespace minicore
